src/java.base/share/classes/jdk/internal/lang/stable/StableFieldUpdaterGenerator.java

```diff
@@ -0,0 +1,347 @@
+                return v == 0 ? applySlowPath(t) : v;
+            } catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        public int applySlowPath(Object t) {
+            preventReentry(this);
+            int v;
+            synchronized (this) {
+                v = (int) accessor.getAcquire(t);
+                if (v == 0) {
+                    try {
+                        v = (int) underlying.invokeExact(t);
+                    } catch (Throwable e) {
+                        throw new RuntimeException(e);
+                    }
+                    accessor.setRelease(t, v);
+                }
+            }
+            return v;
+        }
+    }
+
+    static final MethodHandle LONG_ONCE_APPLY =
+            MhUtil.findVirtual(LOCAL_LOOKUP, LongOnce.class, "apply",
+                    MethodType.methodType(long.class, Object.class));
+
+    record LongOnce(VarHandle accessor,
+                    MethodHandle initialAccessor,
+                    MethodHandle underlying) {
+
+        // Used reflectively
+        @ForceInline
+        public long apply(Object t) {
+            try {
+                long v = (long) initialAccessor.invoke(t);
+                return v == 0 ? applySlowPath(t) : v;
+            } catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        public long applySlowPath(Object t) {
+            preventReentry(this);
+            long v;
+            synchronized (this) {
+                v = (long) accessor.getAcquire(t);
+                if (v == 0) {
+                    try {
+                        v = (long) underlying.invokeExact(t);
+                    } catch (Throwable e) {
+                        throw new RuntimeException(e);
+                    }
+                    accessor.setRelease(t, v);
+                }
+            }
+            return v;
+        }
+
+    }
+
+    static final MethodHandle REFERENCE_ONCE_APPLY =
+            MhUtil.findVirtual(LOCAL_LOOKUP, ReferenceOnce.class, "apply",
+                    MethodType.methodType(Object.class, Object.class));
+
+    record ReferenceOnce(VarHandle accessor,
+                         MethodHandle initialAccessor,
+                         MethodHandle underlying) {
+
+        // Used reflectively
+        @ForceInline
+        public Object apply(Object t) {
+            try {
+                final Object v = (Object) initialAccessor.invoke(t);
+                return v == null ? applySlowPath(t) : v;
+            } catch (Throwable e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        public Object applySlowPath(Object t) {
+            preventReentry(this);
+            Object v;
+            synchronized (this) {
+                v = (Object) accessor.getAcquire(t);
+                if (v == null) {
+                    try {
+                        v = (int) underlying.invokeExact(t);
```

Comment:
Can’t `invokeExact` here, as the return type of `underlying` isn’t guaranteed to be `Object`:
```suggestion
                        v = (Object) underlying.invoke(t);
```